### PR TITLE
Main functionality of CI

### PR DESCRIPTION
**Pull request by tigran10**
I want some feedback, as I am getting close to finishing the code.

* Code goes up to `UAT`. We need to think how to implement another control/gate to let it go until `PROD`. Want to get a feedback first
* I have named all with prefix `ci*`, it is intentional so that i can compare myself later on changes that can be ported back to existing k8s*  functions.  As i did not want to touch them. So its just for me to compare things.
* If the code been committed to a master branch, the `minor` version will be bumped up.
* If the code been commited to a hotfix branch, the `revision` version will be bumped up.
* user will be prompted to own release version, to have a chance to bump up the `major` version if needed
* charts on `feature` branches will be deleted automatically, unless developer want to keep it running, he/she will be prompted to decide `DELETE` or `NOT` with a `timeout`.
* charts on `master` branches will be running in parallel, we need to discuss if this is good or not.
* Deployment to `UAT`, is treated as a `RELEASE` phase. It's almost the same as clicking the release button. It will deploy to UAT, test, and then do release operations.
* Release number is based on a git tag only.
* Release number is in sync between docker and git. Ex, on master It starts as `RC-1.0.0-b1` but later becomes `1.0.0` when deployed to `UAT`. Appropriate git tag will be created.
* `hotfix-*` and `master` branches are generating release candidates and can be released to UAT.
* ...and more  :) 
* Release notes are generated using [gren](https://github.com/github-tools/github-release-notes)

Lets talk :)

### CODE REVIEW

File: CiJenkinsfile [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+                ciPrettyBuildNumber()
+
+                container('git') {
+                    ciWithGitKey(params.rsaKey) {
+                        ciBuildEnvVars()
+                    }
+                }
+
+                container('docker') {
+                    //for feature branch
```
Comment (tigran10) [inline CiJenkinsfile:85]: delete comment

File: CiJenkinsfile [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+                                }
+
+                                container('gren') {
+                                    //https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/
+                                    withCredentials([string(credentialsId: params.githubToken, variable: 'TOKEN')]) {
+                                        sh "gren release --token=${TOKEN}"
+                                    }
+                                }
+
+// my local cm is in a bad state :)
```
Comment (tigran10) [inline CiJenkinsfile]: uncomment for testing

File: file_for_change [hunk 3]
```diff
@@ -0,0 +1,2 @@
+cool feature
```
Comment (tigran10) [inline file_for_change]: delete

File: helm/go-demo-3/templates/deployment.yaml [hunk 4]
```diff
@@ -21,7 +21,7 @@ spec:
     spec:
       containers:
       - name: api
-        image: "vfarcic/go-demo-3:{{ .Values.image.tag }}"
+        image: "digitalinside/go-demo-3:{{ .Values.image.tag }}"
```
Comment (tigran10) [inline helm/go-demo-3/templates/deployment.yaml:24]: put back

File: CiJenkinsfile [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+                                }
+
+                                container('gren') {
+                                    //https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/
+                                    withCredentials([string(credentialsId: params.githubToken, variable: 'TOKEN')]) {
+                                        sh "gren release --token=${TOKEN}"
+                                    }
+                                }
+
+                                container("helm") {
```
Comment (tigran10) [inline CiJenkinsfile:159]: release now ends on pushing chart to helm

File: DeploymentJenkinsfile [hunk 6]
```diff
@@ -0,0 +1,170 @@
+pipeline {
```
Comment (tigran10) [inline DeploymentJenkinsfile:1]: this is separate job, which deploys microservices from  `platform_deployment.yml`

**Comment by vfarcic** [comment]
> Code goes up to UAT. We need to think how to implement another control/gate to let it go until PROD. Want to get a feedback first

GitOps? Can it be yet another PR?

> If the code been commited to a hotfix branch, the revision version will be bumped up.
user will be prompted to own release version, to have a chance to bump up the major version if needed
charts on feature branches will be deleted automatically, unless developer want to keep it running, he/she will be prompted to decide DELETE or NOT with a timeout.

How about having a file with the version number? I'd avoid interactions with Jenkins in form of manually running builds with custom parameters.

> charts on master branches will be running in parallel, we need to discuss if this is good or not.

I'm not sure I understand.

> Deployment to UAT, is treated as a RELEASE phase. It's almost the same as clicking the release button. It will deploy to UAT, test, and then do release operations.

PR instead? Maybe a separate repository that defines an environment?


How about we start putting words (text/chapter) around the code? It might be easier to go step by step and discuss each stage.

**Pull request merged**